ethdb/privateapi/ethbackend.go

```diff
@@ -335,6 +356,87 @@ func (s *EthBackendServer) EngineForkChoiceUpdatedV1(ctx context.Context, req *r
 	}, nil
 }
 
+func (s *EthBackendServer) StartProposer() {
+	go func() {
+		atomic.StoreUint32(s.pauseAssemble, 1)
+		for {
+			if atomic.LoadUint32(s.pauseAssemble) == 1 || atomic.LoadUint32(s.waitingForBeaconChain) == 0 {
+				// Wait really little, this is just not to kill the CPU
+				timer := time.NewTimer(20 * time.Millisecond)
+
+				select {
+				case <-timer.C:
+				case <-s.ctx.Done():
+					return
+				}
+				continue
+			}
+			s.mu.Lock()
+			// Go over each payload and re-update them
+			for id := range s.pendingPayloads {
+				// we do not want to make a copy of the payload in the loop because it contains a lock
+				random := gointerfaces.ConvertH256ToHash(s.pendingPayloads[id].Random)
+				coinbase := gointerfaces.ConvertH160toAddress(s.pendingPayloads[id].Coinbase)
+				timestamp := s.pendingPayloads[id].Timestamp
+				// Tell the stage headers to leave space for the write transaction for mining stages
+				s.skipCycleHack <- struct{}{}
+				block, err := s.assemblePayloadPOS(random, coinbase, timestamp)
```

Comment:
Currently there's a problem: `assemblePayloadPOS` may be called multiple times for the same payload ID. Essentially already assembled blocks are disregarded and the build process starts anew. There's no point in doing that.

Reply:
that is expected, what if you have better transactions in the pool? currently we also have constant updates for proof-of-work mining

Reply:
Better transactions in what sense?

Reply:
in terms of better tips or higher priority depending on who runs the node

Reply:
The problem is that we've already spent time executing a bunch of transactions and calculating the new state root. If we throw all that work away and start anew all the time, we might be over 12 seconds and miss the slot. Theoretically we could listen for new transactions for, say, 11 seconds and only then calculate the new state root during the last second. But even then you probably don't want to throw away all previous execution state, especially for big and lucrative transactions. That would penny-wise, pound-poor.
Also, I suspect that the probably of new transactions that are much more lucrative than those already in the pool arriving within 12 seconds is quite low. In any case, let's start with a simple but sensible approach, with  a view that we can add more optimizations later.

Reply:
okay, we can definetely avoid updating blocks for now, that is how i interpreted the spec, perhaps in must be discussed on ETH R&D